requests-for-comments/2023-03-01-Temporal-RFC.md

```diff
@@ -0,0 +1,239 @@
+
+Since this began we’ve done amazing work in keeping the plugin server going. Chief among the improvements is that we split the plugin server up dependent on the workload into pools of resources that are responsible for different categories of plugin.
+
+### The plugin server, a survey
+
+So what all is the plugin server responsible for? There are really two types of plugins or tasks that we operate on the plugin server: Stateful and Stateless. Stateful plugins make side effects when they are called. This means that an API is called or a mutation is made in a db. Stateless plugins only affect the event as it is in flight. A good example of a stateless plugin is GeoIP where we take the ip that is in the event and based on that information and a static GeoIP db we geolocalize using the ip. 
+
+************************Stateful vs Stateless************************
+
+The practical difference between these two is that a stateless plugin generally executes in a deterministic way. This is mainly because it depends only on its inputs to execute. A stateful plugin on the other hand does not execute so deterministically. This is because of the fact that it depends on something outside of the inputs to the function, like an API. As an example let’s think of a plugin that during execution calls out to an api to either mutate or augment the data that is in flight. This is stateful because depending on the time that this plugin runs the results may vary. If this is a plugin that augments data with the response of an API the data returned by the API may change request to request. The api also may fail or stall. In our case most of these APIs are external to us and may fail or degrade at any point. This makes error handling of stateful plugins very important.
+
+- Plugins
+    - Stateless
+        - Transformations
+    - Stateful
+        - Exports
+        - API Calls
+        - Reverse ETLs
+        - Augmenters
+- Webhooks (Stateful)
+
+So what’s the pain point currently?
+
+Our biggest pain point currently is with the stateful plugins. Regardless of whether they are scheduled or realtime because they depend on external resources that can degrade or fail they require sophisticated timeout and retry logic that we don’t have currently. Another pain point is a relative lack of visibility into the system when things fail. The final pain point is that we don’t have the ability to easily compose these plugins together. What we need is a workflow framework. We are halfway through building one.
+
+You might ask yourself, James, if we are halfway through building a workflow engine, then why don’t we just finish it!? That’s a great question. The real reason we should not complete it is that frameworks already exist to do just this and building another one is not our competitive advantage against our competitors. Building one is a distraction from us building out our actual business.
```

Comment:
> "building one is us building out our actual business"

I would compare us building our own workflow engine to us building our own OLAP database. We didn't do the latter, and found success because of our choice. Many companies/projects have crashed and burned trying to do the former, when _really_ they just wanted to run business logic/data processing reliably + scalably, not faff around with distributed systems.

So far the majority of our customers have been small. We still do not process truly **big** data, and are struggling with what we already have. If we'd like to progress properly we need to choose rock solid foundations, that have already been proved at way beyond the scale we currently see.

Reply:
Companies don't starve, they drown.

Reply:
I'm not here to tell anyone that you _can_ do something when you're all saying you can't. :)